File: lib/swagger.ts
export const swaggerSpec = {
  openapi: '3.0.0',
  info: {
    title: 'Game Web App API',
    description: 'API for the comprehensive game web application with real-time game features',
    version: '1.0.0',
    contact: {
      name: 'Game Team',
      url: 'http://localhost:3000',
    },
  },
  servers: [
    {
      url: 'http://localhost:3000/api',
      description: 'Development server',
    },
  ],
  tags: [
    {
      name: 'Rooms',
      description: 'Game room management endpoints',
    },
    {
      name: 'Players',
      description: 'Player management within rooms',
    },
    {
      name: 'Games',
      description: 'Game management within rooms',
    },
  ],
  paths: {
    '/rooms': {
      post: {
        tags: ['Rooms'],
        summary: 'Create a new game room',
        description: 'Create a new game room with configuration',
        requestBody: {
          required: true,
          content: {
            'application/json': {
              schema: {
                type: 'object',
                required: ['name', 'mainColor', 'maxPlayers', 'pointMode', 'createdBy', 'pointFrom', 'pointTo'],
                properties: {
                  name: {
                    type: 'string',
                    example: 'Quiz Night #1',
                    description: 'Room display name',
                  },
                  mainColor: {
                    type: 'string',
                    example: '#3b82f6',
                    description: 'Primary color for the room (hex format)',
                  },
                  colorFrom: {
                    type: 'string',
                    example: '#10b981',
                    description: 'Gradient color start (hex format)',
                  },
                  colorTo: {
                    type: 'string',
                    example: '#1e40af',
                    description: 'Gradient color end (hex format)',
                  },
                  maxPlayers: {
                    type: 'integer',
                    example: 10,
                    minimum: 1,
                    description: 'Maximum number of players allowed',
                  },
                  pointMode: {
                    type: 'integer',
                    enum: [1, 2],
                    example: 1,
                    description: 'Point calculation mode (1: linear, 2: proportional)',
                  },
                  pointFrom: {
                    type: 'integer',
                    example: 0,
                    description: 'Minimum points',
                  },
                  pointTo: {
                    type: 'integer',
                    example: 100,
                    description: 'Maximum points',
                  },
                  createdBy: {
                    type: 'string',
                    format: 'uuid',
                    example: '550e8400-e29b-41d4-a716-446655440000',
                    description: 'Admin/Creator user ID',
                  },
                },
              },
              example: {
                name: 'Quiz Night #1',
                mainColor: '#3b82f6',
                colorFrom: '#10b981',
                colorTo: '#1e40af',
                maxPlayers: 10,
                pointMode: 1,
                pointFrom: 0,
                pointTo: 100,
                createdBy: '550e8400-e29b-41d4-a716-446655440000',
              },
            },
          },
        },
        responses: {
          200: {
            description: 'Room created successfully',
            content: {
              'application/json': {
                schema: {
                  type: 'object',
                  properties: {
                    id: { type: 'string', format: 'uuid' },
                    name: { type: 'string' },
                    join_code: { type: 'string' },
                    presentation_code: { type: 'string' },
                    main_color: { type: 'string' },
                    max_players: { type: 'integer' },
                    created_at: { type: 'string', format: 'date-time' },
                  },
                },
              },
            },
          },
          400: { description: 'Invalid request parameters' },
          500: { description: 'Server error' },
        },
      },
    },
    '/rooms/{roomId}': {
      get: {
        tags: ['Rooms'],
        summary: 'Get room details',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        responses: {
          200: {
            description: 'Room details retrieved successfully',
            content: {
              'application/json': {
                schema: {
                  type: 'object',
                  properties: {
                    id: { type: 'string', format: 'uuid' },
                    name: { type: 'string' },
                    join_code: { type: 'string' },
                    presentation_code: { type: 'string' },
                    main_color: { type: 'string' },
                    max_players: { type: 'integer' },
                    created_at: { type: 'string', format: 'date-time' },
                  },
                },
              },
            },
          },
          404: { description: 'Room not found' },
        },
      },
    },
    '/rooms/{roomId}/players': {
      get: {
        tags: ['Players'],
        summary: 'Get all players in a room',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        responses: {
          200: {
            description: 'Players list retrieved successfully',
            content: {
              'application/json': {
                schema: {
                  type: 'array',
                  items: {
                    type: 'object',
                    properties: {
                      id: { type: 'string', format: 'uuid' },
                      name: { type: 'string' },
                      score: { type: 'integer' },
                      rank: { type: 'integer' },
                      joined_at: { type: 'string', format: 'date-time' },
                    },
                  },
                },
              },
            },
          },
          404: { description: 'Room not found' },
        },
      },
      post: {
        tags: ['Players'],
        summary: 'Add a player to a room',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        requestBody: {
          required: true,
          content: {
            'application/json': {
              schema: {
                type: 'object',
                required: ['name'],
                properties: {
                  name: {
                    type: 'string',
                    example: 'John Doe',
                  },
                },
              },
            },
          },
        },
        responses: {
          201: {
            description: 'Player added successfully',
            content: {
              'application/json': {
                schema: {
                  type: 'object',
                  properties: {
                    id: { type: 'string', format: 'uuid' },
                    name: { type: 'string' },
                    score: { type: 'integer' },
                    rank: { type: 'integer' },
                  },
                },
              },
            },
          },
          400: { description: 'Invalid request or room at capacity' },
          404: { description: 'Room not found' },
        },
      },
    },
    '/rooms/{roomId}/players/{playerId}': {
      get: {
        tags: ['Players'],
        summary: 'Get player details',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
          {
            name: 'playerId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        responses: {
          200: {
            description: 'Player details retrieved',
            content: {
              'application/json': {
                schema: {
                  type: 'object',
                  properties: {
                    id: { type: 'string', format: 'uuid' },
                    name: { type: 'string' },
                    score: { type: 'integer' },
                    rank: { type: 'integer' },
                  },
                },
              },
            },
          },
          404: { description: 'Player not found' },
        },
      },
      patch: {
        tags: ['Players'],
        summary: 'Update player data',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
          {
            name: 'playerId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        requestBody: {
          required: true,
          content: {
            'application/json': {
              schema: {
                type: 'object',
                properties: {
                  name: { type: 'string' },
                  score: { type: 'integer' },
                  rank: { type: 'integer' },
                },
              },
              example: { score: 100, rank: 1 },
            },
          },
        },
        responses: {
          200: { description: 'Player updated successfully' },
          404: { description: 'Player not found' },
        },
      },
      delete: {
        tags: ['Players'],
        summary: 'Delete a player',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
          {
            name: 'playerId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        responses: {
          200: { description: 'Player deleted successfully' },
          404: { description: 'Player not found' },
        },
      },
    },
    '/rooms/{roomId}/games': {
      get: {
        tags: ['Games'],
        summary: 'Get all games in a room',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        responses: {
          200: {
            description: 'Games list retrieved successfully',
            content: {
              'application/json': {
                schema: {
                  type: 'array',
                  items: {
                    type: 'object',
                    properties: {
                      id: { type: 'string', format: 'uuid' },
                      name: { type: 'string' },
                      type: { type: 'string', enum: ['weight', 'random'] },
                      status: { type: 'string', enum: ['pending', 'active', 'completed'] },
                      game_order: { type: 'integer' },
                      created_at: { type: 'string', format: 'date-time' },
                    },
                  },
                },
              },
            },
          },
          404: { description: 'Room not found' },
        },
      },
      post: {
        tags: ['Games'],
        summary: 'Create a new game',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        requestBody: {
          required: true,
          content: {
            'application/json': {
              schema: {
                type: 'object',
                required: ['type'],
                properties: {
                  type: {
                    type: 'string',
                    enum: ['weight', 'random'],
                    example: 'weight',
                  },
                  name: {
                    type: 'string',
                    example: 'Guessing Game Round 1',
                  },
                  settings: {
                    type: 'object',
                    example: { mode: 'test' },
                  },
                },
              },
            },
          },
        },
        responses: {
          201: {
            description: 'Game created successfully',
            content: {
              'application/json': {
                schema: {
                  type: 'object',
                  properties: {
                    id: { type: 'string', format: 'uuid' },
                    name: { type: 'string' },
                    type: { type: 'string' },
                    status: { type: 'string' },
                    game_order: { type: 'integer' },
                  },
                },
              },
            },
          },
          400: { description: 'Invalid game type' },
          404: { description: 'Room not found' },
        },
      },
    },
    '/rooms/{roomId}/games/{gameId}': {
      get: {
        tags: ['Games'],
        summary: 'Get game details',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
          {
            name: 'gameId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        responses: {
          200: {
            description: 'Game details retrieved',
            content: {
              'application/json': {
                schema: {
                  type: 'object',
                  properties: {
                    id: { type: 'string', format: 'uuid' },
                    name: { type: 'string' },
                    type: { type: 'string' },
                    status: { type: 'string' },
                    settings: { type: 'object' },
                  },
                },
              },
            },
          },
          404: { description: 'Game not found' },
        },
      },
      patch: {
        tags: ['Games'],
        summary: 'Update game status',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
          {
            name: 'gameId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        requestBody: {
          required: true,
          content: {
            'application/json': {
              schema: {
                type: 'object',
                properties: {
                  status: {
                    type: 'string',
                    enum: ['pending', 'active', 'completed'],
                    example: 'active',
                  },
                  settings: { type: 'object' },
                },
              },
            },
          },
        },
        responses: {
          200: { description: 'Game updated successfully' },
          404: { description: 'Game not found' },
        },
      },
      delete: {
        tags: ['Games'],
        summary: 'Delete a game',
        parameters: [
          {
            name: 'roomId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
          {
            name: 'gameId',
            in: 'path',
            required: true,
            schema: { type: 'string', format: 'uuid' },
          },
        ],
        responses: {
          200: { description: 'Game deleted successfully' },
          404: { description: 'Game not found' },
        },
      },
    },
  },
  components: {
    schemas: {
      Room: {
        type: 'object',
        properties: {
          id: { type: 'string', format: 'uuid' },
          name: { type: 'string' },
          join_code: { type: 'string' },
          presentation_code: { type: 'string' },
          main_color: { type: 'string' },
          max_players: { type: 'integer' },
          point_mode: { type: 'integer' },
          created_at: { type: 'string', format: 'date-time' },
        },
      },
      Player: {
        type: 'object',
        properties: {
          id: { type: 'string', format: 'uuid' },
          name: { type: 'string' },
          score: { type: 'integer' },
          rank: { type: 'integer' },
          joined_at: { type: 'string', format: 'date-time' },
        },
      },
      Game: {
        type: 'object',
        properties: {
          id: { type: 'string', format: 'uuid' },
          name: { type: 'string' },
          type: { type: 'string', enum: ['weight', 'random'] },
          status: { type: 'string', enum: ['pending', 'active', 'completed'] },
          game_order: { type: 'integer' },
          settings: { type: 'object' },
          created_at: { type: 'string', format: 'date-time' },
        },
      },
    },
  },
};

File: API/actions/swagger.go
package actions

import (
	"net/http"

	"github.com/gobuffalo/buffalo"
)

// swaggerSpec is the OpenAPI 3.0 specification for this API.
var swaggerSpec = map[string]any{
	"openapi": "3.0.0",
	"info": map[string]any{
		"title":       "Qiz API",
		"version":     "1.0.0",
		"description": "Game room management API",
	},
	"servers": []map[string]any{
		{"url": "http://localhost:3001/api"},
	},
	"paths": map[string]any{
		"/rooms": map[string]any{
			"get": map[string]any{
				"summary":     "List all game rooms",
				"operationId": "listRooms",
				"tags":        []string{"rooms"},
				"responses": map[string]any{
					"200": map[string]any{"description": "OK", "content": jsonArray("GameRoom")},
				},
			},
			"post": map[string]any{
				"summary":     "Create a game room",
				"operationId": "createRoom",
				"tags":        []string{"rooms"},
				"requestBody": jsonBody("CreateRoomRequest"),
				"responses": map[string]any{
					"200": map[string]any{"description": "Created", "content": jsonRef("GameRoom")},
				},
			},
		},
		"/rooms/{room_id}": map[string]any{
			"parameters": []map[string]any{pathParam("room_id")},
			"get": map[string]any{
				"summary": "Get a game room", "operationId": "getRoom", "tags": []string{"rooms"},
				"responses": map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("GameRoom")}},
			},
			"patch": map[string]any{
				"summary": "Update a game room", "operationId": "patchRoom", "tags": []string{"rooms"},
				"requestBody": jsonBody("PatchRoomRequest"),
				"responses":   map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("GameRoom")}},
			},
		},
		"/rooms/{room_id}/players": map[string]any{
			"parameters": []map[string]any{pathParam("room_id")},
			"get": map[string]any{
				"summary": "List players", "operationId": "listPlayers", "tags": []string{"players"},
				"responses": map[string]any{"200": map[string]any{"description": "OK", "content": jsonArray("Player")}},
			},
			"post": map[string]any{
				"summary": "Add a player", "operationId": "createPlayer", "tags": []string{"players"},
				"requestBody": jsonBody("CreatePlayerRequest"),
				"responses":   map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("Player")}},
			},
		},
		"/rooms/{room_id}/players/{player_id}": map[string]any{
			"parameters": []map[string]any{pathParam("room_id"), pathParam("player_id")},
			"get": map[string]any{
				"summary": "Get a player", "operationId": "getPlayer", "tags": []string{"players"},
				"responses": map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("Player")}},
			},
			"patch": map[string]any{
				"summary": "Update a player", "operationId": "patchPlayer", "tags": []string{"players"},
				"requestBody": jsonBody("PatchPlayerRequest"),
				"responses":   map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("Player")}},
			},
			"delete": map[string]any{
				"summary": "Delete a player", "operationId": "deletePlayer", "tags": []string{"players"},
				"responses": map[string]any{"200": map[string]any{"description": "OK"}},
			},
		},
		"/rooms/{room_id}/games": map[string]any{
			"parameters": []map[string]any{pathParam("room_id")},
			"get": map[string]any{
				"summary": "List games", "operationId": "listGames", "tags": []string{"games"},
				"responses": map[string]any{"200": map[string]any{"description": "OK", "content": jsonArray("Game")}},
			},
			"post": map[string]any{
				"summary": "Create a game", "operationId": "createGame", "tags": []string{"games"},
				"requestBody": jsonBody("CreateGameRequest"),
				"responses":   map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("Game")}},
			},
		},
		"/rooms/{room_id}/games/{game_id}": map[string]any{
			"parameters": []map[string]any{pathParam("room_id"), pathParam("game_id")},
			"get": map[string]any{
				"summary": "Get a game", "operationId": "getGame", "tags": []string{"games"},
				"responses": map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("Game")}},
			},
			"patch": map[string]any{
				"summary": "Update a game", "operationId": "patchGame", "tags": []string{"games"},
				"requestBody": jsonBody("PatchGameRequest"),
				"responses":   map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("Game")}},
			},
			"delete": map[string]any{
				"summary": "Delete a game", "operationId": "deleteGame", "tags": []string{"games"},
				"responses": map[string]any{"200": map[string]any{"description": "OK"}},
			},
		},
		"/rooms/{room_id}/games/{game_id}/results": map[string]any{
			"parameters": []map[string]any{pathParam("room_id"), pathParam("game_id")},
			"get": map[string]any{
				"summary": "List results", "operationId": "listResults", "tags": []string{"results"},
				"responses": map[string]any{"200": map[string]any{"description": "OK", "content": jsonArray("GameResult")}},
			},
			"post": map[string]any{
				"summary": "Submit a result", "operationId": "createResult", "tags": []string{"results"},
				"requestBody": jsonBody("CreateResultRequest"),
				"responses":   map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("GameResult")}},
			},
			"put": map[string]any{
				"summary": "Batch submit results", "operationId": "batchResults", "tags": []string{"results"},
				"requestBody": jsonBody("BatchResultsRequest"),
				"responses":   map[string]any{"200": map[string]any{"description": "OK", "content": jsonArray("GameResult")}},
			},
		},
		"/rooms/{room_id}/games/{game_id}/random/winner": map[string]any{
			"parameters": []map[string]any{pathParam("room_id"), pathParam("game_id")},
			"post": map[string]any{
				"summary": "Record random winner", "operationId": "createRandomWinner", "tags": []string{"random"},
				"requestBody": jsonBody("CreateRandomWinnerRequest"),
				"responses":   map[string]any{"200": map[string]any{"description": "OK", "content": jsonRef("RandomWinner")}},
			},
		},
		"/rooms/{room_id}/games/{game_id}/random/winners": map[string]any{
			"parameters": []map[string]any{pathParam("room_id"), pathParam("game_id")},
			"get": map[string]any{
				"summary": "List random winners", "operationId": "listRandomWinners", "tags": []string{"random"},
				"responses": map[string]any{"200": map[string]any{"description": "OK", "content": jsonArray("RandomWinner")}},
			},
		},
	},
	"components": map[string]any{
		"schemas": map[string]any{
			"GameRoom": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"id":                map[string]any{"type": "string", "format": "uuid"},
					"name":              map[string]any{"type": "string"},
					"room_number":       map[string]any{"type": "integer"},
					"join_code":         map[string]any{"type": "string"},
					"presentation_code": map[string]any{"type": "string"},
					"main_color":        map[string]any{"type": "string"},
					"color_from":        map[string]any{"type": "string"},
					"color_to":          map[string]any{"type": "string"},
					"max_players":       map[string]any{"type": "integer"},
					"point_mode":        map[string]any{"type": "string"},
					"point_from":        map[string]any{"type": "integer"},
					"point_to":          map[string]any{"type": "integer"},
					"created_at":        map[string]any{"type": "string", "format": "date-time"},
					"created_by":        map[string]any{"type": "string"},
					"status":            map[string]any{"type": "string"},
					"settings":          map[string]any{"type": "object"},
				},
			},
			"Player": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"id":              map[string]any{"type": "string", "format": "uuid"},
					"room_id":         map[string]any{"type": "string"},
					"name":            map[string]any{"type": "string"},
					"sequence_number": map[string]any{"type": "integer"},
					"joined_at":       map[string]any{"type": "string", "format": "date-time"},
					"score":           map[string]any{"type": "integer"},
					"rank":            map[string]any{"type": "integer"},
					"is_hidden_rank":  map[string]any{"type": "boolean"},
					"is_hidden_score": map[string]any{"type": "boolean"},
					"color":           map[string]any{"type": "string"},
					"metadata":        map[string]any{"type": "object"},
				},
			},
			"Game": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"id":         map[string]any{"type": "string", "format": "uuid"},
					"room_id":    map[string]any{"type": "string"},
					"name":       map[string]any{"type": "string"},
					"type":       map[string]any{"type": "string", "enum": []string{"weight", "random"}},
					"game_order": map[string]any{"type": "integer"},
					"status":     map[string]any{"type": "string"},
					"created_at": map[string]any{"type": "string", "format": "date-time"},
					"settings":   map[string]any{"type": "object"},
				},
			},
			"GameResult": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"id":            map[string]any{"type": "string", "format": "uuid"},
					"game_id":       map[string]any{"type": "string"},
					"player_id":     map[string]any{"type": "string"},
					"points_earned": map[string]any{"type": "integer"},
					"rank":          map[string]any{"type": "integer"},
					"created_at":    map[string]any{"type": "string", "format": "date-time"},
					"player_name":   map[string]any{"type": "string"},
				},
			},
			"RandomWinner": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"id":             map[string]any{"type": "string", "format": "uuid"},
					"game_id":        map[string]any{"type": "string"},
					"player_id":      map[string]any{"type": "string"},
					"admin_action":   map[string]any{"type": "string"},
					"points_awarded": map[string]any{"type": "integer"},
					"created_at":     map[string]any{"type": "string", "format": "date-time"},
				},
			},
			"CreateRoomRequest": map[string]any{
				"type": "object", "required": []string{"name"},
				"properties": map[string]any{
					"name":              map[string]any{"type": "string"},
					"mainColor":         map[string]any{"type": "string"},
					"colorFrom":         map[string]any{"type": "string"},
					"colorTo":           map[string]any{"type": "string"},
					"maxPlayers":        map[string]any{"type": "integer"},
					"pointMode":         map[string]any{"type": "string"},
					"pointFrom":         map[string]any{"type": "integer"},
					"pointTo":           map[string]any{"type": "integer"},
					"createdBy":         map[string]any{"type": "string"},
					"join_code":         map[string]any{"type": "string"},
					"presentation_code": map[string]any{"type": "string"},
				},
			},
			"PatchRoomRequest": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"mainColor": map[string]any{"type": "string"},
					"colorFrom": map[string]any{"type": "string"},
					"colorTo":   map[string]any{"type": "string"},
				},
			},
			"CreatePlayerRequest": map[string]any{
				"type": "object", "required": []string{"name"},
				"properties": map[string]any{"name": map[string]any{"type": "string"}},
			},
			"PatchPlayerRequest": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"name":          map[string]any{"type": "string"},
					"score":         map[string]any{"type": "integer"},
					"isScoreHidden": map[string]any{"type": "boolean"},
					"isRankHidden":  map[string]any{"type": "boolean"},
				},
			},
			"CreateGameRequest": map[string]any{
				"type": "object", "required": []string{"type"},
				"properties": map[string]any{
					"name":     map[string]any{"type": "string"},
					"type":     map[string]any{"type": "string", "enum": []string{"weight", "random"}},
					"settings": map[string]any{"type": "object"},
				},
			},
			"PatchGameRequest": map[string]any{
				"type": "object",
				"properties": map[string]any{
					"status":   map[string]any{"type": "string"},
					"config":   map[string]any{"type": "object"},
					"settings": map[string]any{"type": "object"},
				},
			},
			"CreateResultRequest": map[string]any{
				"type": "object", "required": []string{"playerId", "pointsEarned"},
				"properties": map[string]any{
					"playerId":     map[string]any{"type": "string"},
					"pointsEarned": map[string]any{"type": "integer"},
					"rank":         map[string]any{"type": "integer"},
				},
			},
			"BatchResultsRequest": map[string]any{
				"type": "object", "required": []string{"results"},
				"properties": map[string]any{
					"results": map[string]any{
						"type": "array",
						"items": map[string]any{
							"type": "object",
							"properties": map[string]any{
								"playerId":     map[string]any{"type": "string"},
								"pointsEarned": map[string]any{"type": "integer"},
								"rank":         map[string]any{"type": "integer"},
							},
						},
					},
				},
			},
			"CreateRandomWinnerRequest": map[string]any{
				"type": "object", "required": []string{"playerId", "action"},
				"properties": map[string]any{
					"playerId":      map[string]any{"type": "string"},
					"action":        map[string]any{"type": "string"},
					"pointsAwarded": map[string]any{"type": "integer"},
				},
			},
		},
	},
}

// SwaggerJSONHandler GET /api/swagger.json
func SwaggerJSONHandler(c buffalo.Context) error {
	return c.Render(http.StatusOK, r.JSON(swaggerSpec))
}

// --- tiny helpers for building the spec ---

func jsonRef(schema string) map[string]any {
	return map[string]any{
		"application/json": map[string]any{
			"schema": map[string]any{"$ref": "#/components/schemas/" + schema},
		},
	}
}

func jsonArray(schema string) map[string]any {
	return map[string]any{
		"application/json": map[string]any{
			"schema": map[string]any{
				"type":  "array",
				"items": map[string]any{"$ref": "#/components/schemas/" + schema},
			},
		},
	}
}

func jsonBody(schema string) map[string]any {
	return map[string]any{"required": true, "content": jsonRef(schema)}
}

func pathParam(name string) map[string]any {
	return map[string]any{"name": name, "in": "path", "required": true, "schema": map[string]any{"type": "string"}}
}
